test/test_exportbuild_pkgconfig/CMakeLists.txt

```
@@ -0,0 +1,27 @@
+cmake_minimum_required(VERSION 3.9)
+project(GinkgoExportBuildWithPkgConfigTest LANGUAGES CXX)
+
+find_package(PkgConfig REQUIRED)
+# find_package(Ginkgo REQUIRED)
+pkg_check_modules(GINKGO REQUIRED ginkgo)
+
+# Here, we use test install without any data. We instantiate the
+# interface only.
+message("GINKGOCMAKE ${CMAKE_APPBUNDLE_PATH} ${CMAKE_PREFIX_PATH} ${CMAKE_FRAMEWORK_PATH}")
+message("GINKGO ${GINKGO_PREFIX}")
+message("GINKGO_LIBRARIES ${GINKGO_LIBRARIES}")
+message("GINKGO_INCLUDE_DIRS ${GINKGO_INCLUDE_DIRS}")
+message("GINKGO_CFLAGS ${GINKGO_CFLAGS}")
+message("GINKGO_LDFLAGS ${GINKGO_LDFLAGS}")
+add_executable(test_exportbuild_pkgconfig ../test_install/test_install.cpp)
+target_link_libraries(test_exportbuild_pkgconfig PRIVATE ${GINKGO_LIBRARIES})
+target_link_libraries(test_exportbuild_pkgconfig PRIVATE ${GINKGO_LDFLAGS})
+target_include_directories(test_exportbuild_pkgconfig PRIVATE ${GINKGO_INCLUDE_DIRS})
+target_compile_features(test_exportbuild_pkgconfig PUBLIC cxx_std_14)
+target_link_libraries(test_exportbuild_pkgconfig PRIVATE ${GINKGO_CFLAGS})
```

the pkgconfig support already provides CMake target support. We should probably also add the `-std=c++14` flag to `cflags`
```suggestion
target_link_libraries(test_exportbuild_pkgconfig PRIVATE PkgConfig::GINKGO)
```

---

I think using package config is more close to the original way not the CMake way

---

I would prefer the CMake approach, since that way we can't accidentally miss any part of the pkg-config settings.

---

Although using the target approach make make adds `-pthread` to compile flag, it changes the linking order to lead failure.
```
/usr/bin/c++ /usr/lib/libhwloc.so /usr/lib/libpapi.so -Wl,-rpath -Wl,/usr/lib -Wl,--enable-new-dtags -pthread \
/usr/lib/libmpi.so CMakeFiles/test_exportbuild_pkgconfig.dir/root/ginkgo/test/test_install/test_install.cpp.o -o \
test_exportbuild_pkgconfig  -Wl,-rpath,/usr/local/lib /usr/local/lib/libginkgo.so /usr/local/lib/libginkgo_omp.so \
/usr/local/lib/libginkgo_cuda.so /usr/local/lib/libginkgo_reference.so /usr/local/lib/libginkgo_hip.so \
/usr/local/lib/libginkgo_dpcpp.so /usr/local/lib/libginkgo_device.so 
```
leads `/usr/bin/ld: CMakeFiles/test_exportbuild_pkgconfig.dir/root/ginkgo/test/test_install/test_install.cpp.o: undefined reference to symbol 'papi_sde_init'`

ginkgo.pc
```
-L/usr/local/lib -lginkgo /usr/lib/libhwloc.so -lginkgo_omp -lginkgo_cuda -lginkgo_reference -lginkgo_hip \
-lginkgo_dpcpp -lginkgo_device /usr/lib/libpapi.so -Wl,-rpath -Wl,/usr/lib -Wl,--enable-new-dtags -pthread \
/usr/lib/libmpi.so
```
it seems to put the lib with absolute path like hwloc, papi, mpi before the object file.
However, they should be after object file.
Do you know any documentation or keyword for this behavior?

---

I think the correct way to do things would be to output `-L/usr/lib -lhwloc -lpapi -lmpi`. Since pkg-config is Linux-only, we can rely on libraries being prefixed `lib` with a `.so` or `.a` suffix. I can't find any pkg-config file on my system that uses absolute paths

---

I do not think it is the requirement from the png-config. cmake also change -L -l to the absolute path in the end. pkg-config may prefer we have hwloc.pc, papi.pc, mpi.pc, and any pkg-config file from the outside or third-party package. pkg-config is also supported on Windows, but I can not make it work in this pr

---

pkg-config has few explicit limitations, but on my entire system, I couldn't find a pkg-config file that uses absolute library paths directly, so we are definitely doing something unusual here. CMake's pkg-config support relies on that implicit assumption, treating every -l and -L parameter as library information and all other parameters as link options. This causes the order to be lost.
On the Windows side: We can use `CMAKE_SHARED_LIBRARY_SUFFIX` and friends to make it portable.